### PR TITLE
implement RecHit SOA and move to new framework

### PR DESCRIPTION
**Pull request by VinInn**
in this PR (on top of #312 and #318
TrackingRecHit is introduced as CUDAFormats
RecHit producer migrated to new framework #100 
Clients migrated.
tbd: rename and clean

all three wf tested

also moved to constant memory in doublet builder following Hackaton investigation
(10% speed up in the kernel)

### CODE REVIEW

**Comment by VinInn** [comment]
use
https://github.com/VinInn/cmssw/compare/gpuSmartAllocDoublets...VinInn:gpuNewRecHits
to review the changes introduced by this PR

File: RecoLocalTracker/SiPixelRecHits/src/PixelCPEFast.cc [hunk 1]
```diff
@@ -68,13 +70,19 @@ PixelCPEFast::PixelCPEFast(edm::ParameterSet const & conf,
 
 const pixelCPEforGPU::ParamsOnGPU *PixelCPEFast::getGPUProductAsync(cuda::stream_t<>& cudaStream) const {
   const auto& data = gpuData_.dataForCurrentDeviceAsync(cudaStream, [this](GPUData& data, cuda::stream_t<>& stream) {
+
+      std::cout << "coping pixelCPEforGPU" << std::endl;
+      //here or above???
```
Comment (makortel) [inline RecoLocalTracker/SiPixelRecHits/src/PixelCPEFast.cc]: If you want it to print out when the transfer is initiated, this is the correct place. ("above" outside of the lambda would print every time the product is asked from)

File: CUDADataFormats/TrackingRecHit/src/TrackingRecHit2DCUDA.cc [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  view->m_xg = get32(4);
+  view->m_yg = get32(5);
+  view->m_zg = get32(6);
+  view->m_rg = get32(7);
+
+  m_iphi = view->m_iphi = (int16_t *)get16(0);
+
+  view->m_charge = (int32_t *)get32(8);
+  view->m_xsize = (int16_t *)get16(2);
+  view->m_ysize = (int16_t *)get16(3);
```
[the remark this answers is not in the document]
Comment (VinInn) [reply CUDADataFormats/TrackingRecHit/src/TrackingRecHit2DCUDA.cc]: what difference it makes?

File: RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforGPU.h [hunk 3]
```diff
@@ -9,6 +9,7 @@
 #include "DataFormats/GeometrySurface/interface/SOARotation.h"
 #include "Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h"
 #include "HeterogeneousCore/CUDAUtilities/interface/cuda_cxx17.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaCompat.h"
```
[the remark this answers is not in the document]
Comment (VinInn) [reply RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforGPU.h]: not strictly needed at the moment...

File: RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitFromSOA.cc [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  descriptions.add("siPixelRecHitFromSOA",desc);
+}
+
+
+void SiPixelRecHitFromSOA::acquire(edm::Event const& iEvent, edm::EventSetup const& iSetup, edm::WaitingTaskWithArenaHolder waitingTaskHolder) {
+
+  CUDAProduct<TrackingRecHit2DCUDA> const& inputDataWrapped = iEvent.get(tokenHit_);
+
+  // try to be in parallel with tracking
+  CUDAScopedContext ctx{iEvent.streamID(), std::move(waitingTaskHolder)};
```
[the remark this answers is not in the document]
Comment (VinInn) [reply RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitFromSOA.cc]: ok

File: RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitHeterogeneous.cc [hunk 5]
```diff
@@ -29,15 +30,9 @@
 
 #include <cuda_runtime.h>
 
-class SiPixelRecHitHeterogeneous: public HeterogeneousEDProducer<heterogeneous::HeterogeneousDevices <
-                                                                   heterogeneous::GPUCuda,
-                                                                   heterogeneous::CPU
-                                                                   > > {
+class SiPixelRecHitHeterogeneous : public edm::global::EDProducer<> {
```
[the remark this answers is not in the document]
Comment (VinInn) [reply RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitHeterogeneous.cc]: Indeed, in the plan

File: RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitHeterogeneous.cc [hunk 6]
```diff
@@ -173,142 +97,30 @@ void SiPixelRecHitHeterogeneous::acquireGPUCuda(const edm::HeterogeneousEvent& i
...
-    // - to set up edm::Refs to the clusters
-    edm::Handle<SiPixelClusterCollectionNew> hclusters;
-    iEvent.getByToken(clusterToken_, hclusters);
+  assert(cpe);
+  assert(clusters.clusModuleStart());
+  TrackingRecHit2DCUDA hits(nHits, cpe, clusters.clusModuleStart(), ctx.stream());
 
-    convertGPUtoCPU(iEvent.event(), hclusters, *output);
-  }
+  gpuAlgo_.makeHitsAsync(hits,digis, clusters, bs, fcpe->getGPUProductAsync(ctx.stream()), ctx.stream());
```
[the remark this answers is not in the document]
Comment (VinInn) [reply RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitHeterogeneous.cc]: originally I though to keep separate the CPU class with storage in the producer and have the algo only depending on the View. This did not work out as one needs more pointers on the host.
So yes, it is a possibility

File: RecoLocalTracker/SiPixelRecHits/plugins/gpuPixelRecHits.h [hunk 7]
```diff
@@ -118,34 +115,52 @@ namespace gpuPixelRecHits {
...
 
+
+
+                          int32_t * chargeh = &hits.charge(0);
+                          uint16_t * detInd = &hits.detectorIndex(0);
+                          float * xg = &hits.xGlobal(0); float * yg = &hits.yGlobal(0); float * zg = &hits.zGlobal(0); float * rg = &hits.rGlobal(0);
+                          int16_t * iph = &hits.iphi(0);
+                          float * xl = &hits.xLocal(0); float * yl = &hits.yLocal(0);
+                          float * xe = &hits.xerrLocal(0); float * ye = &hits.yerrLocal(0);
+                          int16_t * xs = &hits.clusterSizeX(0); int16_t * ys = &hits.clusterSizeY(0);
```
[the remark this answers is not in the document]
Comment (VinInn) [reply RecoLocalTracker/SiPixelRecHits/plugins/gpuPixelRecHits.h]: yeah, was just copied from the signature.
will fix and beatify

File: RecoLocalTracker/SiPixelRecHits/python/SiPixelRecHits_cfi.py [hunk 8]
```diff
@@ -6,6 +6,40 @@
...
+    cpu = _siPixelRecHitsPreSplitting.clone()
+)
+
+from Configuration.ProcessModifiers.gpu_cff import gpu
+gpu.toModify(siPixelRecHitsPreSplitting,
+    cuda = cms.EDAlias(
+        siPixelRecHitsLegacyPreSplitting = cms.VPSet(
+            cms.PSet(type = cms.string("SiPixelRecHitedmNewDetSetVector"))
+        )
+    )
```
[the remark this answers is not in the document]
Comment (VinInn) [reply RecoLocalTracker/SiPixelRecHits/python/SiPixelRecHits_cfi.py]: Ok, thanks. I just tried to make it working waiting for your explanation of the meaning of all that

File: RecoLocalTracker/SiPixelRecHits/src/PixelCPEFast.cc (hunk 1 above)
Comment (VinInn) [reply RecoLocalTracker/SiPixelRecHits/src/PixelCPEFast.cc]: yes, indeed. I wanted just to make sure it was really done once
(I was planning to fill here some __constant__ stuff, but does not work across libraries)

File: CUDADataFormats/TrackingRecHit/src/TrackingRecHit2DCUDA.cc [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+  // if empy do not bother
+  if (0==nHits) {
+    cudautils::copyAsync(m_view, view, stream);
+    return;
+  }
+
+
+  m_store16 = cs->make_device_unique<uint16_t[]>(nHits*n16,stream);
+  m_store32 = cs->make_device_unique<float[]>(nHits*n32+11+(1+TrackingRecHit2DSOAView::Hist::wsSize())/sizeof(float),stream);
```
[the remark this answers is not in the document]
Comment (VinInn) [reply CUDADataFormats/TrackingRecHit/src/TrackingRecHit2DCUDA.cc]: right, I was thinking to introduce a stride function computing the required stride
 ((n*b+127)/128)*128/b  and use it as stride(nHits,4) stride(nHIts,2) consistently.
it is also true that these arrays are accessed mostly random, so does not make much difference if aligned

File: RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforGPU.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (VinInn) [reply RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforGPU.h]: actually is needed due to the device functions....

File: CUDADataFormats/TrackingRecHit/src/TrackingRecHit2DCUDA.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (makortel) [reply CUDADataFormats/TrackingRecHit/src/TrackingRecHit2DCUDA.cc]: Aesthetics (also code rules)

File: CUDADataFormats/TrackingRecHit/src/TrackingRecHit2DCUDA.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (makortel) [reply CUDADataFormats/TrackingRecHit/src/TrackingRecHit2DCUDA.cc]: Ok.

File: RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforGPU.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (makortel) [reply RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforGPU.h]: So the header gets included in some CPU .cc file as well? Ok.

[the remark this answers is not in the document]
Comment (VinInn) [reply RecoLocalTracker/SiPixelRecHits/interface/pixelCPEforGPU.h]: yes, there is the possibility of using this very CPE in the standard CPU wfs...

**Comment by fwyzard** [comment]
This PR should be superseded by #324 / #329, apart from https://github.com/cms-patatrack/cmssw/pull/324#issuecomment-487865923 .

**Pull request closed without merging**